File: Euclid/Euclid.cs
using Euclid.Util;
using log4net;
using log4net.Config;
using System;
using System.IO;
using System.Reflection;
using Euclid.Storage.Database;
using Euclid.Network;
using Euclid.Storage.Database.Access;
using Euclid.Game;
using Euclid.Messages;
using System.Threading;
using System.Linq;
using Euclid.Network.Streams;
using System.Collections.Generic;

namespace Euclid
{
    class Euclid
    {
        #region Fields

        private static readonly ILog log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
        public static readonly bool Encryption = true;

        #endregion

        #region Properties

        /// <summary>
        /// Get the logger instance.
        /// </summary>
        public static ILog Logger
        {
            get { return log; }
        }

        /// <summary>
        /// Get the official release supported
        /// </summary>
        public static string ClientVersion
        {
            get { return "RELEASE63-201302211227-193109692"; }
        }

        #endregion

        static void Main(string[] args)
        {
            var logRepository = LogManager.GetRepository(Assembly.GetEntryAssembly());
            XmlConfigurator.Configure(logRepository, new FileInfo("log4net.config"));

            Console.Title = "Euclid - Habbo Hotel Emulation";

            log.Info("Booting Euclid - Written by Quackster");
            log.Info("Emulation of Habbo Hotel 2007 Shockwave client");

            /*
            var encrypt_rc4 = new RC4();
            var key = encrypt_rc4.CreateKey();
            encrypt_rc4.SetKey(key);
            var encipher = encrypt_rc4.Encipher("test");

            var decrypt_rc4 = new RC4();
            decrypt_rc4.SetKey(key);
            var decipher = decrypt_rc4.Decipher(encipher);
            Console.WriteLine(encipher + " == " + decipher);
            */

            /*
            var rc4 = new RC4();
            rc4.SetKey(1337);
            Console.WriteLine(rc4.Decipher("6055992B"));
            // Console.WriteLine(rc4.Encipher("test"));
            */

            try
            {
                tryDatabaseConnection();
                tryGameData();
                tryCreateServer();

                log.Info("Server has started!");

                while (true) Thread.Sleep(100);
            }
            catch (Exception ex)
            {
                log.Error(ex);
            }

#if DEBUG
            Console.Read();
#endif

        }

        #region Private methods

        /// <summary>
        /// Test database connection
        /// </summary>
        private static void tryDatabaseConnection()
        {
            log.Info("Attempting to connect to database");
            SessionFactoryBuilder.Instance.InitialiseSessionFactory(ServerConfig.Instance.ConnectionString);
            log.Info("Connection successful!");
        }

        /// <summary>
        /// Load game data
        /// </summary>
        private static void tryGameData()
        {
            RoomDao.ResetVisitorCounts();

            PermissionsManager.Instance.Load();
            ValueManager.Instance.Load();
            ItemManager.Instance.Load();
            CatalogueManager.Instance.Load();
            RoomManager.Instance.Load();
            NavigatorManager.Instance.Load();
            MessageHandler.Instance.Load();
            PluginManager.Instance.Load();        
        }

        /// <summary>
        /// Try and create server
        /// </summary>
        private static void tryCreateServer()
        {
            GameServer.Logger.Info("Starting server");

            var roomIds = ServerConfig.Instance.ConfigValues
                .Where(x => x.Key.Contains("server/public/port/"))
                .Select(x => x.Key.Replace("server/public/port/", ""))
                .ToList();

            var gameAddresses = new List<GameAddress>();
            
            foreach (var roomId in roomIds)
            {
                gameAddresses.Add(new GameAddress(ServerConfig.Instance.GetString("server", "public", "ip", roomId), ServerConfig.Instance.GetInt("server", "public", "port", roomId), int.Parse(roomId)));
            }

            GameServer.CreateServer(
                new GameAddress(ServerConfig.Instance.GetString("server", "main", "ip"), ServerConfig.Instance.GetInt("server", "main", "port")),
                new GameAddress(ServerConfig.Instance.GetString("server", "private", "ip"), ServerConfig.Instance.GetInt("server", "private", "port")),
                gameAddresses
            );

            GameServer.Instance.InitialiseServer();

            foreach (var gameAddress in gameAddresses)
            {
                var roomData = RoomDao.GetRoomData(gameAddress.RoomId);

                if (roomData == null)
                    continue;

                GameServer.Logger.Info($"[{roomData.Name}] is listening on port: {gameAddress.IpAddress}:{gameAddress.Port}!");
            }

            GameServer.Logger.Info($"Private server is now listening on port: {GameServer.Instance.PrivateServer.IpAddress}:{GameServer.Instance.PrivateServer.Port}!");
            GameServer.Logger.Info($"Main server is now listening on port: {GameServer.Instance.MainServer.IpAddress}:{GameServer.Instance.MainServer.Port}!");
        }

        public class Game
        {
        }

        #endregion
    }
}

File: euclid/src/main/java/euclid/Euclid.java
package euclid;

import euclid.game.catalogue.CatalogueManager;
import euclid.game.item.ItemManager;
import euclid.game.navigator.NavigatorManager;
import euclid.game.permissions.PermissionsManager;
import euclid.game.plugins.PluginManager;
import euclid.game.room.RoomManager;
import euclid.game.values.ValueManager;
import euclid.messages.MessageHandler;
import euclid.network.GameServer;
import euclid.network.streams.GameAddress;
import euclid.storage.database.SessionFactoryBuilder;
import euclid.storage.database.access.RoomDao;
import euclid.storage.database.data.RoomData;
import euclid.util.ServerConfig;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.core.config.Configurator;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public class Euclid {

    private static final Logger log = LogManager.getLogger(Euclid.class);

    public static final boolean ENCRYPTION = true;

    public static final String CLIENT_VERSION = "RELEASE63-201302211227-193109692";

    public static Logger getLogger() {
        return log;
    }

    public static void main(String[] args) {
        initialiseLogging();
        setConsoleTitle("Euclid - Habbo Hotel Emulation");

        log.info("Booting Euclid - Written by Quackster");
        log.info("Emulation of Habbo Hotel 2007 Shockwave client");

        try {
            tryDatabaseConnection();
            tryGameData();
            tryCreateServer();

            log.info("Server has started!");

            while (true) {
                Thread.sleep(100);
            }
        } catch (Exception ex) {
            log.error("Server failed to start", ex);
        }
    }

    private static void initialiseLogging() {
        for (String path : new String[] { "config/log4j2.xml", "log4j2.xml" }) {
            if (Files.exists(Path.of(path))) {
                Configurator.initialize((String) null, path);
                return;
            }
        }
    }

    private static void setConsoleTitle(String title) {
        System.out.print("\u001b]0;" + title + "\u0007");
        System.out.flush();
    }

    private static void tryDatabaseConnection() {
        log.info("Attempting to connect to database");
        SessionFactoryBuilder.getInstance().initialiseSessionFactory(ServerConfig.getInstance().getConnectionString());
        log.info("Connection successful!");
    }

    private static void tryGameData() {
        RoomDao.resetVisitorCounts();

        PermissionsManager.getInstance().load();
        ValueManager.getInstance().load();
        ItemManager.getInstance().load();
        CatalogueManager.getInstance().load();
        RoomManager.getInstance().load();
        NavigatorManager.getInstance().load();
        MessageHandler.getInstance().load();
        PluginManager.getInstance().load();
    }

    private static void tryCreateServer() {
        log.info("Starting server");

        ServerConfig config = ServerConfig.getInstance();
        List<String> roomIds = new ArrayList<>();
        for (String key : config.getConfigValues().keySet()) {
            if (key.startsWith("server/public/port/")) {
                roomIds.add(key.substring("server/public/port/".length()));
            }
        }

        List<GameAddress> gameAddresses = new ArrayList<>();
        for (String roomId : roomIds) {
            gameAddresses.add(new GameAddress(
                    config.getString("server", "public", "ip", roomId),
                    config.getInt("server", "public", "port", roomId),
                    Integer.parseInt(roomId)
            ));
        }

        GameServer.createServer(
                new GameAddress(config.getString("server", "main", "ip"), config.getInt("server", "main", "port")),
                new GameAddress(config.getString("server", "private", "ip"), config.getInt("server", "private", "port")),
                gameAddresses
        );

        GameServer.getInstance().initialiseServer();

        for (GameAddress gameAddress : gameAddresses) {
            RoomData roomData = RoomDao.getRoomData(gameAddress.getRoomId());
            if (roomData == null) {
                continue;
            }
            log.info("[{}] is listening on port: {}:{}!", roomData.getName(), gameAddress.getIpAddress(), gameAddress.getPort());
        }

        GameServer mainServer = GameServer.getInstance();
        log.info("Private server is now listening on port: {}:{}!", mainServer.getPrivateServer().getIpAddress(), mainServer.getPrivateServer().getPort());
        log.info("Main server is now listening on port: {}:{}!", mainServer.getMainServer().getIpAddress(), mainServer.getMainServer().getPort());
    }

    public static final class Game {
    }
}
